### PR TITLE
feat: implement recursive challenger

### LINKED ISSUE #57: Link inputs, challenge wires and the hash table
Instead of having a Challenger table which would consist only of CTLs, we can write a circuit to link the input proof and public values to the hash table. 

Before each challenge sampling, we gather all the observed inputs, pad to fit a multiple of the hash input size, and send them to the hash table (optionally, if we forego the Sponge table, then we would additionally need to cut the batch of values into chunks of size the hash input size and send those chunks to the hash table instead).

### PR DESCRIPTION
**Pull request by Nashtare**
## Summary

Implements a recursive `CircuitChallenger` that is compatible with the native Plonky3 `DuplexChallenger`.

Below is a list of changes but I've added a section in the `mdbook` with more details on the approach to give a better grasp as the changes are consequent.

Fixes #57

## Changes
- Coefficient-level state management to match native challenger's base field operations
- Support for both D=1 (base field) and D=4 (extension field) Poseidon2 configurations
- Recompose/decompose operations to bridge coefficient state with extension field permutations (this is done on the _same_ Poseidon2 trace table, and incurs for the base-field case an overhead of ~70 extra rows split over the different primitive tables.

## Tests
- [x] Unit tests added/updated
- [x] Existing tests pass
- [x] Edge cases considered (briefly note any)

## Benchmarks / Performance (if relevant)
- [ ] Before/after numbers included
- [ ] Benchmark command or link to CI perf output
- [ ] No unexpected regressions in CPU / memory / proof size / compile time

## Docs (if relevant)
- [x] Rustdoc updated
- [x] Book/README updated
- [ ] Examples updated

## Risk

There are still no CTLs involved to ensure proper mapping, this will be done later on.
Also genericity of Poseidon2 perm (currently hardcoded to D=4, W=16 for 32 bit fields).

### CODE REVIEW

File: book/src/hashing.md [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+
+```
+┌────────────────────────────────────────────────────────────────┐
+│                     Duplex Sponge Operation                    │
+├────────────────────────────────────────────────────────────────┤
+│                                                                │
+│   observe(value):                                              │
+│     1. Clear output_buffer (any pending outputs are invalid)   │
+│     2. Push value to input_buffer                              │
+│     3. If input_buffer.len() == RATE:                          │
```
Comment (4l0n50) [inline book/src/hashing.md]: and if not it only does only 1 and 2?

**Review by 4l0n50 — COMMENTED** [review]
I did a first pass. I would like to understand what is preventing calling directly `add_poseidon2_perm` for constraining the duplexing input and output wires.

File: recursion/src/challenger/circuit.rs [hunk 2]
```diff
@@ -1,115 +1,322 @@
...
+                .recompose_base_coeffs_to_ext::<BF>(
+                    &self.state[3 * EF::DIMENSION..4 * EF::DIMENSION],
+                )
+                .expect("recomposition should succeed"),
+        ];
+
+        // 3. Apply Poseidon2 permutation via witness hint
+        // (CTLs for soundness can be added later)
+        let ext_outputs = circuit
+            .add_poseidon2_perm_for_challenger(self.poseidon2_config, ext_inputs)
```
[the remark this answers is not in the document]
Comment (Nashtare) [reply recursion/src/challenger/circuit.rs]: `add_poseidon2_perm()` is tailored for MMCS and 2-to-1 compression, and does not expose all the output limbs meanwhile this method (and the _base) equivalent is tailored for absorption / squeezing.

We could unify them together but I thought it would complicate a bit the interface on the MMCS side so I left `add_poseidon2_perm` untouched.
